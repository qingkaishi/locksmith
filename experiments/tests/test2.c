#include <pthread.h>
#include <stdio.h>

int f(pthread_mutex_t *l) {
  int k, *j;
  
  pthread_mutex_init(l, NULL);
  pthread_mutex_lock(l);

  j = &k;
  if(pthread_mutex_trylock(l)) {
    *j = 5;
  }
  else {
    j = (int *) &l;
  }
  pthread_mutex_unlock(l);
  return 2;
}

int g() {
  pthread_mutex_t l;
  f(&l);
  return 0;
}

void * dostuff(void *);

void * dostuff(void * a) {
  pthread_mutex_t t;
  printf("foo");
  *(int *)a = 0;
  f(&t);
}
int main() {
  int i = 1;
	pthread_t t1;

  pthread_create(&t1, NULL, dostuff, &i);
  pthread_create(&t1, NULL, dostuff, &i);
	if (i) {
		i = 0;
	}
  i = 0;
}


