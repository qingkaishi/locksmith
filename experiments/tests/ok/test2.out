************************* STARTING *************************


test2.c:15: Warning: test2.c:15: Probably using char* where void* is best used:
 lock(*l:test2.c:4)*^{l:test2.c:4}*^{&l:test2.c:4} flows to int*^{j:test2.c:5}.  Losing precision.


test2.c:15: Warning: assigning incompatible pointers, conflating flow

test2.c:15: Warning: lock is conflated! (*l:test2.c:4)

test2.c:15: Warning: assigning number to pointer
functions declared and used but not defined:
  printf
shared: &i:test2.c:36

Warning: Possible data race: &i:test2.c:36 is not protected!
 references:
  dereference of cast:test2.c:32 at test2.c:32
    &i:test2.c:36 => cast:test2.c:32 at dostuff test2.c:39
  locks acquired:
    <empty>
  in: main at test2.c:35 

  dereference of &i:test2.c:36 at test2.c:41
    &i:test2.c:36
  locks acquired:
    <empty>
  in: main at test2.c:35 

  dereference of &i:test2.c:36 at test2.c:42
    &i:test2.c:36
  locks acquired:
    <empty>
  in: main at test2.c:35 

  dereference of &i:test2.c:36 at test2.c:44
    &i:test2.c:36
  locks acquired:
    <empty>
  in: main at test2.c:35 


*************************** DONE ***************************
